consensus/src/dag/dag_fetcher.rs

```diff
@@ -163,21 +166,34 @@ impl RpcHandler for FetchRequestHandler {
     fn process(&mut self, message: Self::Request) -> anyhow::Result<Self::Response> {
         let dag_reader = self.dag.read();
 
+        // `Certified Node`: In the good case, there should exist at least one honest validator that
+        // signed the Certified Node that has the all the parents to fulfil this
+        // request.
+        // `Node`: In the good case, the sender of the Node should have the parents in its local DAG
+        // to satisfy this request.
         ensure!(
-            message
-                .parents()
-                .iter()
-                .all(|metadata| dag_reader.get_node(metadata).is_some()),
+            dag_reader.all_exists(message.targets().iter()),
             FetchRequestHandleError::ParentsMissing
         );
 
         let certified_nodes: Vec<_> = dag_reader
-            .reachable_from_parents(
-                message.parents(),
+            .reachable(
+                message.targets(),
                 Some(message.exists_bitmask().first_round()),
                 |_| true,
             )
-            .map(|node_status| node_status.as_node().clone().deref().clone())
+            .filter_map(|node_status| {
+                let arc_node = node_status.as_node();
+                self.author_to_index
+                    .get(arc_node.author())
+                    .and_then(|author_idx| {
+                        if !message.exists_bitmask().has(arc_node.round(), *author_idx) {
+                            Some(arc_node.as_ref().clone())
+                        } else {
+                            None
+                        }
+                    })
+            })
```

Comment:
Can you add the filter_map as an argument to reachable?

Reply:
I think that would mean I need to move self.author_to_index and that means I would need to clone it. 

Or I need to make reachable resolve the author to index and pass it to the filter. 

Not sure whichever is readable.